Different overlap checking between `__mul__` and `__rmul__`

**Bug Report**

When attempting to add overrides to the `__mul__` and `__rmul__` magic methods, different behaviour is observed for the same set of overrides.

This seems similar to the issue in #10755, but that has since been fixed.

**To Reproduce**

Playground link: https://mypy-play.net/?mypy=latest&python=3.12&gist=1297febcc3b38a2eea422e7073842c31

(Note that this playground is for the latest version (1.14.1), but the same errors are observed on `master`)

Sample program:

```py3
from typing import TypeVar, overload
from decimal import Decimal

TAmount = TypeVar("TAmount", int, Decimal)


class Unit:
    @overload
    def __mul__(self, other: Unit) -> Unit: ...
    @overload
    def __mul__(self, other: TAmount) -> Quantity[TAmount]: ...

    def __mul__(self, other: Unit | TAmount) -> Unit | Quantity[TAmount]:
        raise NotImplementedError

    @overload
    def __rmul__(self, other: Unit) -> Unit: ...
    @overload
    def __rmul__(self, other: TAmount) -> Quantity[TAmount]: ...

    def __rmul__(self, other: Unit | TAmount) -> Unit | Quantity[TAmount]:
        raise NotImplementedError


class Quantity[TAmount]:
    def __init__(self, value: TAmount, units: Unit) -> None:
        self.value = value
        self.units = units
```

**Expected Behavior**

I expect this program to typecheck successfull (or error for both `__mul__` and `__rmul__`)

**Actual Behavior**

```
main.py:22: error: Signatures of "__rmul__" of "Unit" and "__mul__" of "Unit | int" are unsafely overlapping  [misc]
main.py:22: error: Signatures of "__rmul__" of "Unit" and "__mul__" of "Unit | Decimal" are unsafely overlapping  [misc]
Found 2 errors in 1 file (checked 1 source file)
```

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -95,6 +95,7 @@ from mypy.nodes import (
     OperatorAssignmentStmt,
     OpExpr,
     OverloadedFuncDef,
+    OverloadPart,
     PassStmt,
     PromoteExpr,
     RaiseStmt,
@@ -399,6 +400,11 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
         # argument through various `checker` and `checkmember` functions.
         self._is_final_def = False
 
+        # Track when we enter an overload implementation. Some checks should not be applied
+        # to the implementation signature when specific overloads are available.
+        # Use `enter_overload_impl` to modify.
+        self.overload_impl_stack: list[OverloadPart] = []
+
         # This flag is set when we run type-check or attribute access check for the purpose
         # of giving a note on possibly missing "await". It is used to avoid infinite recursion.
         self.checking_missing_await = False
@@ -660,7 +666,8 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
         if num_abstract not in (0, len(defn.items)):
             self.fail(message_registry.INCONSISTENT_ABSTRACT_OVERLOAD, defn)
         if defn.impl:
-            defn.impl.accept(self)
+            with self.enter_overload_impl(defn.impl):
+                defn.impl.accept(self)
         if not defn.is_property:
             self.check_overlapping_overloads(defn)
             if defn.type is None:
@@ -683,6 +690,14 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                 self.msg.no_overridable_method(defn.name, defn)
             self.check_explicit_override_decorator(defn, found_method_base_classes, defn.impl)
             self.check_inplace_operator_method(defn)
+
+    @contextmanager
+    def enter_overload_impl(self, impl: OverloadPart) -> Iterator[None]:
+        self.overload_impl_stack.append(impl)
+        try:
+            yield
+        finally:
+            assert self.overload_impl_stack.pop() == impl
 
     def extract_callable_type(self, inner_type: Type | None, ctx: Context) -> CallableType | None:
         """Get type as seen by an overload item caller."""
@@ -1203,7 +1218,11 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                     )
 
                 if name:  # Special method names
-                    if defn.info and self.is_reverse_op_method(name):
+                    if (
+                        defn.info
+                        and self.is_reverse_op_method(name)
+                        and defn not in self.overload_impl_stack
+                    ):
                         self.check_reverse_op_method(item, typ, name, defn)
                     elif name in ("__getattr__", "__getattribute__"):
                         self.check_getattr_method(typ, defn, name)

--- a/test-data/unit/check-classes.test
+++ b/test-data/unit/check-classes.test
@@ -2487,6 +2487,60 @@ reveal_type(Num3() + Num1())  # N: Revealed type is "__main__.Num3"
 reveal_type(Num2() + Num3())  # N: Revealed type is "__main__.Num2"
 reveal_type(Num3() + Num2())  # N: Revealed type is "__main__.Num3"
 
+[case testReverseOperatorWithOverloads3]
+from typing import Union, overload
+
+class A:
+    def __mul__(self, value: A, /) -> A: ...
+    def __rmul__(self, value: A, /) -> A: ...
+
+class B:
+    @overload
+    def __mul__(self, other: B, /) -> B: ...
+    @overload
+    def __mul__(self, other: A, /) -> str: ...
+    def __mul__(self, other: Union[B, A], /) -> Union[B, str]: pass
+
+    @overload
+    def __rmul__(self, other: B, /) -> B: ...
+    @overload
+    def __rmul__(self, other: A, /) -> str: ...
+    def __rmul__(self, other: Union[B, A], /) -> Union[B, str]: pass
+
+[case testReverseOperatorWithOverloadsNested]
+from typing import Union, overload
+
+class A:
+    def __mul__(self, value: A, /) -> A: ...
+    def __rmul__(self, value: A, /) -> A: ...
+
+class B:
+    @overload
+    def __mul__(self, other: B, /) -> B: ...
+    @overload
+    def __mul__(self, other: A, /) -> str: ...
+    def __mul__(self, other: Union[B, A], /) -> Union[B, str]: pass
+
+    @overload
+    def __rmul__(self, other: B, /) -> B: ...
+    @overload
+    def __rmul__(self, other: A, /) -> str: ...
+    def __rmul__(self, other: Union[B, A], /) -> Union[B, str]:
+        class A1:
+            def __add__(self, other: C1) -> int: ...
+
+        class B1:
+            def __add__(self, other: C1) -> int: ...
+
+        class C1:
+            @overload
+            def __radd__(self, other: A1) -> str: ...   # E: Signatures of "__radd__" of "C1" and "__add__" of "A1" are unsafely overlapping
+            @overload
+            def __radd__(self, other: B1) -> str: ...   # E: Signatures of "__radd__" of "C1" and "__add__" of "B1" are unsafely overlapping
+            def __radd__(self, other): pass
+
+        return ""
+
 [case testDivReverseOperator]
 # No error: __div__ has no special meaning in Python 3
 class A1:

